MCOL-5594: Interactive "mcs cluster stop" command for CMAPI.

mariadb-AlanMologorsky opened the pull request:
- [x] [add] NodeProcessController class to handle Node operations
- [x] [add] two endpoints: stop_dmlproc (PUT) and is_process_running (GET)
- [x] [add] NodeProcessController.put_stop_dmlproc method to separately stop DMLProc on primary Node
- [x] [add] NodeProcessController.get_process_running method to check if specified process running or not
- [x] [add] build_url function to helpers.py. It needed to build urls with query_params
- [x] [add] MCSProcessManager.gracefully_stop_dmlproc method
- [x] [add] MCSProcessManager.is_service_running method as a top level wrapper to the same method in dispatcher
- [x] [fix] MCSProcessManager.stop by using new gracefully_stop_dmlproc
- [x] [add] interactive option and mode to mcs cluster stop command
- [x] [fix] requirements.txt with typer version to 0.9.0 where supports various of features including "Annotated"
- [x] [fix] requirements.txt click version (8.1.3 -> 8.1.7) and typing-extensions (4.3.0 -> 4.8.0). This is dependencies for typer package. 
- [x] [fix] multiple minor formatting, docstrings and comments
- [x] [add] TransactionManager ContextDecorator to manage transactions in less code and in one place
- [x] [add] TransactionManager to cli cluster stop command and to API cluster shutdown command
- [x] [fix] id -> txn_id in ClusterHandler class
- [x] [fix] ClusterHandler.shutdown class to use inside existing transaction
- [x] [add] docstrings in multiple places

mariadb-AlanMologorsky committed:
MCOL-5594: Interactive "mcs cluster stop" command for CMAPI.

[add] NodeProcessController class to handle Node operations
[add] two endpoints: stop_dmlproc (PUT) and is_process_running (GET)
[add] NodeProcessController.put_stop_dmlproc method to separately stop DMLProc on primary Node
[add] NodeProcessController.get_process_running method to check if specified process running or not
[add] build_url function to helpers.py. It needed to build urls with query_params
[add] MCSProcessManager.gracefully_stop_dmlproc method
[add] MCSProcessManager.is_service_running method as a top level wrapper to the same method in dispatcher
[fix] MCSProcessManager.stop by using new gracefully_stop_dmlproc
[add] interactive option and mode to mcs cluster stop command
[fix] requirements.txt with typer version to 0.9.0 where supports various of features including "Annotated"
[fix] requirements.txt click version (8.1.3 -> 8.1.7) and typing-extensions (4.3.0 -> 4.8.0). This is dependencies for typer package.
[fix] multiple minor formatting, docstrings and comments

drrtuy commented on cmapi/cmapi_server/handlers/cluster.py:
```diff
@@ -141,7 +141,8 @@ def start(
         config: str = DEFAULT_MCS_CONF_PATH,
-        logger: logging.Logger = logging.getLogger('cmapi_server')
+        logger: logging.Logger = logging.getLogger('cmapi_server'),
+        timeout: int = 10
```
Units used are unclear to me.

drrtuy replied:
10 seconds looks too short as a timeout for shutdown.

mariadb-AlanMologorsky replied:
Changed to 15 seconds. Added units to docstring. Anyway it's only for next releases.

drrtuy reviewed (commented):
I see no issues with the patch itself. I suggest you to ask Allen to look at the printouts wording.

mariadb-AlanMologorsky committed:
MCOL-5594: Add new CMAPI transaction manager.

- [add] TransactionManager ContextDecorator to manage transactions in less code and in one place
- [add] TransactionManager to cli cluster stop command and to API cluster shutdown command
- [fix] id -> txn_id in ClusterHandler class
- [fix] ClusterHandler.shutdown class to use inside existing transaction
- [add] docstrings in multiple places

drrtuy commented on cmapi/cmapi_server/handlers/cluster.py:
```diff
@@ -150,6 +154,11 @@ def shutdown(
         :type logger: logging.Logger, optional
+        :param in_transaction: is function called in existing transaction or no
+        :type in_transaction: bool
+        :param timeout: timeout in seconds for gracefully stop DMLProc
```
`to gracefully stop` or for `a graceful stop` IMHO.

mariadb-AllenHerrera replied:
timeout in seconds to gracefully stop DMLProc

mariadb-AlanMologorsky replied:
Fixed.

mariadb-AllenHerrera commented on cmapi/cmapi_server/managers/process.py:
```diff
@@ -262,6 +262,41 @@ def noop(cls, *args, **kwargs):
+                )
+        except (ConnectionRefusedError, RuntimeError):
+            logging.error(
+                'Cannot set SS_ROLLBACK and SS_SHUTDOWN_PENDING '
```
Ideas:
"Cannot set SS_ROLLBACK and SS_SHUTDOWN_PENDING using DBRM while trying to gracefully auto stop DMLProc."
to
"Cannot set SS_ROLLBACK and SS_SHUTDOWN_PENDING via DBRM, graceful auto stop of DMLProc failed. Try a regular stop method"
or
"Graceful auto stop of DMLProc failed: Cannot set SS_ROLLBACK and SS_SHUTDOWN_PENDING using DBRM. Try a regular stop method"

mariadb-AlanMologorsky replied:
@mariadb-AllenHerrera Try or Trying?

mariadb-AllenHerrera commented on cmapi/cmapi_server/managers/transaction.py:
```diff
@@ -0,0 +1,136 @@
+            self._handle_exception(exc=exc)
+        if not suceeded:
+            self._handle_exception(
+                exc=CMAPIBasicError('Starting transaction isn\'t succesfull.')
```
successful

mariadb-AlanMologorsky replied:
Done.

mariadb-AllenHerrera commented on cmapi/mcs_cluster_tool/cluster_app.py:
```diff
@@ -32,9 +44,119 @@ def status():
+        typer.Option(
+            '--interactive/--no-interactive', '-i/-no-i',
+            help=(
+                'Use this option on a highly loaded cluster. '
```
How about this? if accurate
**Use this option on a active cluster as interactive stop waits for current writes to complete in DMLProc before shutting down. Ensuring consistency, preventing data loss of active writes.**
or shorter
Ensures consistency and prevents data loss by waiting for active writes to complete.

mariadb-AlanMologorsky replied:
First one is marvelous!

@mariadb-AllenHerrera but I'm confused about `a active cluster` not `an active cluster`. Why?

drrtuy replied:
Just remove `a` adjective allows one to drop article.

mariadb-AlanMologorsky replied:
Done.

mariadb-AllenHerrera commented on cmapi/mcs_cluster_tool/cluster_app.py:
```diff
@@ -32,9 +44,119 @@ def status():
+        typer.Option(
+            '-t', '--timeout',
+            help=(
+                'Time in seconds to wait DMLproc gracefully stops. '
```
Ideas:
Time in seconds to wait for DMLproc to gracefully stop. 
Warning: Low wait timeout values could result in data loss if the cluster is very active.

mariadb-AlanMologorsky replied:
Done.

mariadb-AllenHerrera commented on cmapi/mcs_cluster_tool/cluster_app.py:
```diff
@@ -32,9 +44,119 @@ def status():
+        typer.Option(
+            '--force/--no-force', '-f/-no-f',
+            help=(
+                'Force stopping a cluster.'
```
Idea:
Force stops columnstore
Warning: This could cause data corruption and/or data loss

mariadb-AlanMologorsky replied:
Done.

mariadb-AllenHerrera commented on cmapi/mcs_cluster_tool/cluster_app.py:
```diff
@@ -32,9 +44,119 @@ def status():
+                break
+            else:
+                force = typer.confirm(
+                    'DMLProc still working. '
```
Idea
DMLProc is still running.
Do you want to force stop?
WARNING: Could cause data loss and or broken cluster

mariadb-AlanMologorsky replied:
Done.

mariadb-AllenHerrera commented on cmapi/mcs_cluster_tool/cluster_app.py:
```diff
@@ -32,9 +44,119 @@ def status():
+                break
+            else:
+                force = typer.confirm(
+                    'DMLProc still working. '
```
Idea
DMLProc is still running.
Do you want to force stop?
WARNING: Could cause data loss and or broken cluster

mariadb-AlanMologorsky replied:
Duplicate)

mariadb-AlanMologorsky committed:
MCOL-5594: Review fixes.

drrtuy reviewed (approved):
untested ACK
